### ISSUE TITLE
Groupby with function returning a tuple fails if index has a name

### ISSUE THREAD

**Issue by achabotl** [pandas-dev/pandas#18696]
#### Code Sample

```python
import pandas as pd
print(pd.__version__)
idx = pd.date_range('2017-12-08', periods=6, freq='10D')
df = pd.DataFrame(np.arange(6), index=idx)
df.index.name = 'name'
df.groupby(lambda x: (x.year, x.month)).mean()
```

#### Problem description

In 0.21.0 (and 0.20.x), calling `groupy` with a function that returns a tuple on a DataFrame with a named index fails with `ValueError: Names should be list-like for a MultiIndex`. It seems to be caused by the index having a name. 

The sample works in 0.19.0, with the output below. The same code without `df.index.name = 'name'` works under 0.20 and 0.21.

#### Expected Output

```
            0
(2017, 12)  1
(2018, 1)   4
```
#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: None
pip: 9.0.1
setuptools: 38.2.4
Cython: 0.27.3
numpy: 1.13.3
scipy: None
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

**Comment by jreback** [comment]
I believe this is correct, even though it looks a bit weird. You are passing a callable

e.g
```
In [10]: df.groupby([df.index.year, df.index.month]).mean()
Out[10]: 
           0
name name   
2017 12    1
2018 1     4
```

```
In [14]: df.groupby(df.index).mean()
Out[14]: 
            0
name         
2017-12-08  0
2017-12-18  1
2017-12-28  2
2018-01-07  3
2018-01-17  4
2018-01-27  5
```

**Comment by jreback** [comment]
related to #11579

@chris-b1 @TomAugspurger

**Comment by jreback** [comment]
we might just need to doc this

**Comment by jreback** [comment]
@achabotl a couple of examples in the doc-string might be helpful here. I'll mark it as such.

**Comment by VincentLa** [comment]
I can potentially work on this during Pycon Sprints (2018)!

**Comment by leevei** [comment]
Should this be closed since it works in 0.23.4?

**Comment by mbirdi** [comment]
I wrote an example notebook for myself documenting the index-name-groupby behavior in 0.22.0.  I can post that notebook if people think its helpful.

I may need a little help posting the notebook.  I don't use github much and am stuck.  I tried to use nbviewer.jupyter.org to post the notebook, but I guess that website renders the notebook, and does not host it.  So I want to use github to host the notebook and nbviewer to render it, but I must be missing something.
